pkg/storage/s3.go

```diff
@@ -1,56 +1,327 @@
+		return errors.New("access_key not found")
+	}
+	if options.AccessKey != "" && options.SecretAccessKey == "" {
+		return errors.New("secret_access_key not found")
 	}
-	// TODO: Verify Region.
 
 	s3.Endpoint = options.Endpoint
 	s3.Region = options.Region
+	// StorageClass, SSE and ACL are acceptable to be empty
+	s3.StorageClass = options.StorageClass
+	s3.Sse = options.SSE
+	s3.Acl = options.ACL
+	s3.AccessKey = options.AccessKey
+	s3.SecretAccessKey = options.SecretAccessKey
+	s3.ForcePathStyle = options.ForcePathStyle
 	return nil
 }
 
 func defineS3Flags(flags *pflag.FlagSet) {
-	flags.String(s3EndpointOption, "", "Set the AWS S3 endpoint URL")
-	flags.String(s3RegionOption, "", "Set the AWS region")
-	// TODO: Finalize the list of options.
-	_ = flags.MarkHidden(s3EndpointOption)
+	flags.String(s3EndpointOption, "", "Set the S3 endpoint URL, please specify the http or https scheme explicitly")
+	flags.String(s3RegionOption, "", "Set the S3 region, e.g. us-east-1")
+	flags.String(s3StorageClassOption, "", "Set the S3 storage class, e.g. STANDARD")
+	flags.String(s3SSEOption, "", "Set the S3 server-side encryption algorithm, e.g. AES256")
+	flags.String(s3ACLOption, "", "Set the S3 canned ACLs, e.g. authenticated-read")
+	flags.String(s3ProviderOption, "", "Set the S3 provider, e.g. aws, alibaba, ceph")
 }
 
 func getBackendOptionsFromS3Flags(flags *pflag.FlagSet) (options S3BackendOptions, err error) {
+	sendCredential, err = flags.GetBool(flagSendCredentialOption)
+	if err != nil {
+		err = errors.Trace(err)
+		return
+	}
 	options.Endpoint, err = flags.GetString(s3EndpointOption)
 	if err != nil {
 		err = errors.Trace(err)
 		return
 	}
-
 	options.Region, err = flags.GetString(s3RegionOption)
 	if err != nil {
 		err = errors.Trace(err)
 		return
 	}
+	options.SSE, err = flags.GetString(s3SSEOption)
+	if err != nil {
+		err = errors.Trace(err)
+		return
+	}
+	options.ACL, err = flags.GetString(s3ACLOption)
+	if err != nil {
+		err = errors.Trace(err)
+		return
+	}
+	options.StorageClass, err = flags.GetString(s3StorageClassOption)
+	if err != nil {
+		err = errors.Trace(err)
+		return
+	}
+	options.ForcePathStyle = true
+	options.Provider, err = flags.GetString(s3ProviderOption)
+	if err != nil {
+		err = errors.Trace(err)
+		return
+	}
+
+	return options, err
+}
+
+// newS3Storage initialize a new s3 storage for metadata
+func newS3Storage(backend *backup.S3) (*S3Storage, error) {
+	qs := *backend
+	v := credentials.Value{
+		AccessKeyID:     qs.AccessKey,
+		SecretAccessKey: qs.SecretAccessKey,
+	}
 
-	// TODO: Add more options here.
+	// low timeout to ec2 metadata service
+	lowTimeoutClient := &http.Client{Timeout: time.Duration(timeout) * time.Second}
+	def := defaults.Get()
+	def.Config.HTTPClient = lowTimeoutClient
+	ec2Session, err := session.NewSession()
+	if err != nil {
+		return nil, err
+	}
 
-	return
+	// first provider to supply a credential set "wins"
+	providers := []credentials.Provider{
```

Comment:
Adding an explicit chain here can easily mess up the default chain behavior. It seems like we want the default chain unless the credentials are already present in program memory. So this can be changed to:

``` go
if qs.AccessKey != "" {
    aws_config.WithCredentials(credentials.NewStaticCredentials(qs.AccessKey, qs.SecretAccessKey))
} else {
   // Nothing to do here!
}

Reply:
I just add the static value provider before the default providers, it should have no impact to default chain behavior.
Anyway, I have updated following your suggestion, PTAL again, thanks!